File: js/test/kms/rankingApi.test.ts
import process from 'process';

import { MapleStoryApiError, MapleStoryApiErrorCode } from '../../src';
import { MapleStoryApi } from '../../src/maplestory/api/kms';
import { toString } from '../utils';

const apiKey = process.env.API_KEY_KMS!; // Your API Key
const api = new MapleStoryApi(apiKey);

const ocid = 'e0a4f439e53c369866b55297d2f5f4eb';

describe('Ranking Information Retrieval', () => {
  describe('getOverallRanking', () => {
    test('success: getOverallRanking', async () => {
      const response = await api.getOverallRanking();
      console.log(toString(response));
    });

    test('success: getOverallRanking with paging', async () => {
      const response = await api.getOverallRanking({
        page: 2,
      });
      console.log(toString(response));
    });

    test('success: getOverallRanking with ocid', async () => {
      const response = await api.getOverallRanking({
        ocid: ocid,
      });
      console.log(toString(response));
    });

    test('success: getOverallRanking with specific job', async () => {
      const response = await api.getOverallRanking({
        characterClass: '전사-히어로',
      });
      console.log(toString(response));
    });

    test('success: getOverallRanking with job group', async () => {
      const response = await api.getOverallRanking({
        characterClass: '전사-전체 전직',
      });
      console.log(toString(response));
    });

    test('success: getOverallRanking with world name', async () => {
      const response = await api.getOverallRanking({
        worldName: '스카니아',
      });
      console.log(toString(response));
    });

    test('success: getOverallRanking with world type', async () => {
      const response = await api.getOverallRanking({
        worldType: 0,
      });
      console.log(toString(response));
    });

    test('success: getOverallRanking with world and job', async () => {
      const response = await api.getOverallRanking({
        worldName: '스카니아',
        characterClass: '아델-전체 전직'
      });
      console.log(toString(response));
    });

    test('success: getOverallRanking with date', async () => {
      const response = await api.getOverallRanking(undefined, {
        year: 2023,
        month: 12,
        day: 23,
      });
      console.log(toString(response));
    });

    test('fail: getOverallRanking with invalid date', async () => {
      try {
        await api.getOverallRanking(undefined, {
          year: 2023,
          month: 12,
          day: 21,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2023-12-22.');
        console.log(error.message);
      }
    });

    test('fail: getOverallRanking with invalid ocid throw OPENAPI00003', async () => {
      const invalidOcid = 'invalid_ocid_123';
      try {
        await api.getOverallRanking({ ocid: invalidOcid });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getUnionRanking', () => {
    test('success: getUnionRanking', async () => {
      const response = await api.getUnionRanking();
      console.log(toString(response));
    });

    test('success: getUnionRanking with paging', async () => {
      const response = await api.getUnionRanking({
        page: 2,
      });
      console.log(toString(response));
    });

    test('success: getUnionRanking with ocid', async () => {
      const response = await api.getUnionRanking({
        ocid: ocid,
      });
      expect(response.ranking.length).toBe(1);
      console.log(toString(response));
    });

    test('success: getUnionRanking with world name', async () => {
      const response = await api.getUnionRanking({
        worldName: '스카니아',
      });
      expect(response.ranking.length).toBeGreaterThan(0);
      console.log(toString(response));
    });

    test('success: getUnionRanking with date', async () => {
      const response = await api.getUnionRanking(undefined, {
        year: 2023,
        month: 12,
        day: 23,
      });
      console.log(toString(response));
    });

    test('fail: getUnionRanking with invalid date', async () => {
      try {
        await api.getUnionRanking(undefined, {
          year: 2023,
          month: 12,
          day: 21,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2023-12-22.');
        console.log(error.message);
      }
    });

    test('fail: getUnionRanking with invalid ocid throw OPENAPI00003', async () => {
      const invalidOcid = 'invalid_ocid_123';
      try {
        await api.getUnionRanking({ ocid: invalidOcid });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getGuildRanking', () => {
    test('success: getGuildRanking', async () => {
      const response = await api.getGuildRanking();
      console.log(toString(response));
    });

    test('success: getGuildRanking with paging', async () => {
      const response = await api.getGuildRanking({
        rankingType: 0,
        page: 2,
      });
      console.log(toString(response));
    });

    test('success: getGuildRanking with guild name', async () => {
      const response = await api.getGuildRanking({
        rankingType: 0,
        guildName: '리더',
      });
      expect(response.ranking.length).toBeGreaterThan(0);
      console.log(toString(response));
    });

    test('success: getGuildRanking with world name', async () => {
      const response = await api.getGuildRanking({
        rankingType: 0,
        worldName: '스카니아',
      });
      expect(response.ranking.length).toBeGreaterThan(0);
      console.log(toString(response));
    });

    test('success: getGuildRanking with world and guild name', async () => {
      const response = await api.getGuildRanking({
        rankingType: 0,
        worldName: '스카니아',
        guildName: '리더',
      });
      expect(response.ranking.length).toBeLessThanOrEqual(1);
      console.log(toString(response));
    });

    test('success: getGuildRanking with date', async () => {
      const response = await api.getGuildRanking(undefined, {
        year: 2023,
        month: 12,
        day: 23,
      });
      console.log(toString(response));
    });

    test('fail: getGuildRanking with invalid date', async () => {
      try {
        await api.getGuildRanking(undefined, {
          year: 2023,
          month: 12,
          day: 21,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2023-12-22.');
        console.log(error.message);
      }
    });

    test('success: getGuildRanking with invalid guild name', async () => {
      const invalidGuildName = 'invalid_guild_name_123';
      const response = await api.getGuildRanking({
        rankingType: 0,
        guildName: invalidGuildName,
      });
      expect(response.ranking.length).toBe(0);
      console.log(toString(response));
    });

    test('fail: getGuildRanking with invalid ranking type throw OPENAPI00004', async () => {
      try {
        await api.getGuildRanking({
          rankingType: -1,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00004);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getDojangRanking', () => {
    test('success: getDojangRanking', async () => {
      const response = await api.getDojangRanking();
      console.log(toString(response));
    });

    test('success: getDojangRanking with paging', async () => {
      const response = await api.getDojangRanking({
        difficulty: 1,
        page: 2,
      });
      console.log(toString(response));
    });

    test('success: getDojangRanking with ocid', async () => {
      const response = await api.getDojangRanking({
        difficulty: 1,
        ocid: ocid,
      });
      expect(response.ranking.length).toBeGreaterThanOrEqual(0);
      console.log(toString(response));
    });

    test('success: getDojangRanking with world name', async () => {
      const response = await api.getDojangRanking({
        difficulty: 1,
        worldName: '스카니아',
      });
      console.log(toString(response));
    });

    test('success: getDojangRanking with specific job', async () => {
      const response = await api.getDojangRanking({
        difficulty: 1,
        characterClass: '전사-히어로',
      });
      console.log(toString(response));
    });

    test('success: getDojangRanking with date', async () => {
      const response = await api.getDojangRanking(undefined, {
        year: 2023,
        month: 12,
        day: 23,
      });
      console.log(toString(response));
    });

    test('fail: getDojangRanking with invalid date', async () => {
      try {
        await api.getDojangRanking(undefined, {
          year: 2023,
          month: 12,
          day: 21,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2023-12-22.');
        console.log(error.message);
      }
    });

    test('fail: getDojangRanking with invalid ocid throw OPENAPI00003', async () => {
      const invalidOcid = 'invalid_ocid_123';
      try {
        await api.getDojangRanking({ difficulty: 1, ocid: invalidOcid });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getSeedRanking', () => {
    test('success: getSeedRanking', async () => {
      const response = await api.getSeedRanking();
      console.log(toString(response));
    });

    test('success: getSeedRanking with paging', async () => {
      const response = await api.getSeedRanking({
        page: 2,
      });
      console.log(toString(response));
    });

    test('success: getSeedRanking with ocid', async () => {
      const response = await api.getSeedRanking({
        ocid: ocid,
      });
      expect(response.ranking.length).toBeGreaterThanOrEqual(0);
      console.log(toString(response));
    });

    test('success: getSeedRanking with world name', async () => {
      const response = await api.getSeedRanking({
        worldName: '스카니아',
      });
      expect(response.ranking.length).toBeGreaterThan(0);
      console.log(toString(response));
    });

    test('success: getSeedRanking with date', async () => {
      const response = await api.getSeedRanking(undefined, {
        year: 2023,
        month: 12,
        day: 23,
      });
      console.log(toString(response));
    });

    test('fail: getSeedRanking with invalid date', async () => {
      try {
        await api.getSeedRanking(undefined, {
          year: 2023,
          month: 12,
          day: 21,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2023-12-22.');
        console.log(error.message);
      }
    });

    test('fail: getSeedRanking with invalid ocid throw OPENAPI00003', async () => {
      const invalidOcid = 'invalid_ocid_123';
      try {
        await api.getSeedRanking({ ocid: invalidOcid });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getAchievementRanking', () => {
    test('success: getAchievementRanking', async () => {
      const response = await api.getAchievementRanking();
      console.log(toString(response));
    });

    test('success: getAchievementRanking with paging', async () => {
      const response = await api.getAchievementRanking({
        page: 2,
      });
      console.log(toString(response));
    });

    test('success: getAchievementRanking with ocid', async () => {
      const response = await api.getAchievementRanking({
        ocid: ocid,
      });
      expect(response.ranking.length).toBeGreaterThanOrEqual(0);
      console.log(toString(response));
    });

    test('success: getAchievementRanking with date', async () => {
      const response = await api.getAchievementRanking(undefined, {
        year: 2023,
        month: 12,
        day: 23,
      });
      console.log(toString(response));
    });

    test('fail: getAchievementRanking with invalid date', async () => {
      try {
        await api.getAchievementRanking(undefined, {
          year: 2023,
          month: 12,
          day: 21,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2023-12-22.');
        console.log(error.message);
      }
    });

    test('fail: getAchievementRanking with invalid ocid throw OPENAPI00003', async () => {
      const invalidOcid = 'invalid_ocid_123';
      try {
        await api.getAchievementRanking({ ocid: invalidOcid });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });
  });
});

File: python/test/kms/ranking_api_test.py
import os
import sys
import unittest
import pytest

from datetime import datetime
from dotenv import load_dotenv

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../')))

from maplestory_openapi.api.common import MapleStoryApiException
from maplestory_openapi.api.kms import MapleStoryApi

load_dotenv()

api_key = os.environ['API_KEY_KMS'] # Your API Key
api = MapleStoryApi(api_key)

ocid = 'e0a4f439e53c369866b55297d2f5f4eb'

# Ranking Information Retrieval

class TestGetOverallRanking(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_overall_ranking(self):
        response = await api.get_overall_ranking()
        print(response)

    async def test_success_get_overall_ranking_with_paging(self):
        response = await api.get_overall_ranking(page=2)
        print(response)

    async def test_success_get_overall_ranking_with_ocid(self):
        response = await api.get_overall_ranking(ocid=ocid)
        print(response)

    async def test_success_get_overall_ranking_with_specific_job(self):
        response = await api.get_overall_ranking(class_name='전사-히어로')
        print(response)

    async def test_success_get_overall_ranking_with_job_group(self):
        response = await api.get_overall_ranking(class_name='전사-전체 전직')
        print(response)

    async def test_success_get_overall_ranking_with_world_name(self):
        response = await api.get_overall_ranking(world_name='스카니아')
        print(response)

    async def test_success_get_overall_ranking_with_world_type(self):
        response = await api.get_overall_ranking(world_type=0)
        print(response)

    async def test_success_get_overall_ranking_with_world_and_job(self):
        response = await api.get_overall_ranking(world_name='스카니아', class_name='아델-전체 전직')
        print(response)

    async def test_success_get_overall_ranking_with_date(self):
        response = await api.get_overall_ranking(date=datetime(2023, 12, 23))
        print(response)

    async def test_fail_get_overall_ranking_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_overall_ranking(date=datetime(2023, 12, 21))
        assert 'You can only retrieve data after 2023-12-22' in str(e.value)
        print(e.value)

    async def test_fail_get_overall_ranking_with_invalid_ocid_throw_OPENAPI00003(self):
        invalid_ocid = 'invalid_ocid_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_overall_ranking(ocid=invalid_ocid)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)


class TestGetUnionRanking(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_union_ranking(self):
        response = await api.get_union_ranking()
        print(response)

    async def test_success_get_union_ranking_with_paging(self):
        response = await api.get_union_ranking(page=2)
        print(response)

    async def test_success_get_union_ranking_with_ocid(self):
        response = await api.get_union_ranking(ocid=ocid)
        assert len(response.ranking) == 1
        print(response)

    async def test_success_get_union_ranking_with_world_name(self):
        response = await api.get_union_ranking(world_name='스카니아')
        assert len(response.ranking) > 0
        print(response)

    async def test_success_get_union_ranking_with_date(self):
        response = await api.get_union_ranking(date=datetime(2023, 12, 23))
        print(response)

    async def test_fail_get_union_ranking_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_union_ranking(date=datetime(2023, 12, 21))
        assert 'You can only retrieve data after 2023-12-22' in str(e.value)
        print(e.value)

    async def test_fail_get_union_ranking_with_invalid_ocid_throw_OPENAPI00003(self):
        invalid_ocid = 'invalid_ocid_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_union_ranking(ocid=invalid_ocid)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)


class TestGetGuildRanking(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_guild_ranking(self):
        response = await api.get_guild_ranking()
        print(response)

    async def test_success_get_guild_ranking_with_paging(self):
        response = await api.get_guild_ranking(ranking_type=0, page=2)
        print(response)

    async def test_success_get_guild_ranking_with_guild_name(self):
        response = await api.get_guild_ranking(ranking_type=0, guild_name='리더')
        assert len(response.ranking) > 0
        print(response)

    async def test_success_get_guild_ranking_with_world_name(self):
        response = await api.get_guild_ranking(ranking_type=0, world_name='스카니아')
        assert len(response.ranking) > 0
        print(response)

    async def test_success_get_guild_ranking_with_world_and_guild_name(self):
        response = await api.get_guild_ranking(ranking_type=0, world_name='스카니아', guild_name='리더')
        assert len(response.ranking) <= 1
        print(response)

    async def test_success_get_guild_ranking_with_date(self):
        response = await api.get_guild_ranking(date=datetime(2023, 12, 23))
        print(response)

    async def test_fail_get_guild_ranking_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_guild_ranking(date=datetime(2023, 12, 21))
        assert 'You can only retrieve data after 2023-12-22' in str(e.value)
        print(e.value)

    async def test_success_get_guild_ranking_with_invalid_guild_name(self):
        invalid_guild_name = 'invalid_guild_name_123'
        response = await api.get_guild_ranking(ranking_type=0, guild_name=invalid_guild_name)
        assert len(response.ranking) == 0
        print(response)

    async def test_fail_get_guild_ranking_with_invalid_ranking_type_throw_OPENAPI00004(self):
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_guild_ranking(ranking_type=-1)
        assert e.value.error_code == 'OPENAPI00004'
        print(e.value.error_code, e.value.message)


class TestGetDojangRanking(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_dojang_ranking(self):
        response = await api.get_dojang_ranking()
        print(response)

    async def test_success_get_dojang_ranking_with_paging(self):
        response = await api.get_dojang_ranking(difficulty=1, page=2)
        print(response)

    async def test_success_get_dojang_ranking_with_ocid(self):
        response = await api.get_dojang_ranking(difficulty=1, ocid=ocid)
        assert len(response.ranking) >= 0
        print(response)

    async def test_success_get_dojang_ranking_with_world_name(self):
        response = await api.get_dojang_ranking(difficulty=1, world_name='스카니아')
        print(response)

    async def test_success_get_dojang_ranking_with_specific_job(self):
        response = await api.get_dojang_ranking(difficulty=1, class_name='전사-히어로')
        print(response)

    async def test_success_get_dojang_ranking_with_date(self):
        response = await api.get_dojang_ranking(date=datetime(2023, 12, 23))
        print(response)

    async def test_fail_get_dojang_ranking_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_dojang_ranking(date=datetime(2023, 12, 21))
        assert 'You can only retrieve data after 2023-12-22' in str(e.value)
        print(e.value)

    async def test_fail_get_dojang_ranking_with_invalid_ocid_throw_OPENAPI00003(self):
        invalid_ocid = 'invalid_ocid_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_dojang_ranking(difficulty=1, ocid=invalid_ocid)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)


class TestGetTheSeedRanking(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_theseed_ranking(self):
        response = await api.get_theseed_ranking()
        print(response)

    async def test_success_get_theseed_ranking_with_paging(self):
        response = await api.get_theseed_ranking(page=2)
        print(response)

    async def test_success_get_theseed_ranking_with_ocid(self):
        response = await api.get_theseed_ranking(ocid=ocid)
        assert len(response.ranking) >= 0
        print(response)

    async def test_success_get_theseed_ranking_with_world_name(self):
        response = await api.get_theseed_ranking(world_name='스카니아')
        assert len(response.ranking) > 0
        print(response)

    async def test_success_get_theseed_ranking_with_date(self):
        response = await api.get_theseed_ranking(date=datetime(2023, 12, 23))
        print(response)

    async def test_fail_get_theseed_ranking_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_theseed_ranking(date=datetime(2023, 12, 21))
        assert 'You can only retrieve data after 2023-12-22' in str(e.value)
        print(e.value)

    async def test_fail_get_theseed_ranking_with_invalid_ocid_throw_OPENAPI00003(self):
        invalid_ocid = 'invalid_ocid_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_theseed_ranking(ocid=invalid_ocid)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)


class TestGetAchievementRanking(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_achievement_ranking(self):
        response = await api.get_achievement_ranking()
        print(response)

    async def test_success_get_achievement_ranking_with_paging(self):
        response = await api.get_achievement_ranking(page=2)
        print(response)

    async def test_success_get_achievement_ranking_with_ocid(self):
        response = await api.get_achievement_ranking(ocid=ocid)
        assert len(response.ranking) >= 0
        print(response)

    async def test_success_get_achievement_ranking_with_date(self):
        response = await api.get_achievement_ranking(date=datetime(2023, 12, 23))
        print(response)

    async def test_fail_get_achievement_ranking_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_achievement_ranking(date=datetime(2023, 12, 21))
        assert 'You can only retrieve data after 2023-12-22' in str(e.value)
        print(e.value)

    async def test_fail_get_achievement_ranking_with_invalid_ocid_throw_OPENAPI00003(self):
        invalid_ocid = 'invalid_ocid_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_achievement_ranking(ocid=invalid_ocid)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)


if __name__ == '__main__':
    unittest.main()
